File: public/condense_csv.js
#!/usr/bin/env node
/**
 * CSV Condenser Script
 * Removes redundant T2-T5 components from WorkingCopy.csv, keeping only T1 versions.
 * 
 * The T2-T5 versions are predictable since they only add the previous tier as the first ingredient
 * while keeping all other ingredients the same.
 */

const fs = require('fs');
const path = require('path');

class CSVCondenser {
    constructor() {
        this.inputFile = 'finalComponentList_new.csv';
        this.outputFile = 'finalComponentList_condensed.csv';
        this.backupFile = 'finalComponentList_new_backup.csv';
    }

    // Parse CSV row handling quoted fields and commas
    parseCSVRow(row) {
        const result = [];
        let current = '';
        let inQuotes = false;

        for (let i = 0; i < row.length; i++) {
            const char = row[i];

            if (char === '"') {
                inQuotes = !inQuotes;
            } else if (char === ',' && !inQuotes) {
                result.push(current.trim());
                current = '';
            } else {
                current += char;
            }
        }

        result.push(current.trim());
        return result;
    }

    // Convert array back to CSV row
    arrayToCSVRow(arr) {
        return arr.map(field => {
            // Quote fields that contain commas or quotes
            if (field.includes(',') || field.includes('"')) {
                return `"${field.replace(/"/g, '""')}"`;
            }
            return field;
        }).join(',');
    }

    // Process the CSV file
    condenseCSV() {
        try {
            console.log('📖 Reading CSV file...');
            const content = fs.readFileSync(this.inputFile, 'utf8');
            const lines = content.split('\n').filter(line => line.trim());

            console.log(`Found ${lines.length} lines`);

            // Parse header
            const header = this.parseCSVRow(lines[0]);
            console.log('Header columns:', header.length);

            // Process all rows
            const condensedRows = [];
            const processedComponents = new Set();

            for (let i = 1; i < lines.length; i++) {
                const row = this.parseCSVRow(lines[i]);

                if (row.length < header.length) {
                    console.log(`Skipping malformed row ${i}: ${row.length} columns`);
                    continue;
                }

                const outputId = row[0];
                const outputName = row[1];
                const outputTier = parseInt(row[5]) || 1;

                // Only process T1 versions (base recipes)
                if (outputTier === 1) {
                    // Remove tier designation from ID and name
                    const condensedId = outputId.replace(/-t1$/, '');
                    const condensedName = outputName.replace(/ T1$/, '');

                    // Check if we've already processed this component
                    if (processedComponents.has(condensedId)) {
                        console.log(`Duplicate component found: ${condensedId}`);
                        continue;
                    }

                    processedComponents.add(condensedId);

                    // Create condensed row
                    const condensedRow = [...row];
                    condensedRow[0] = condensedId;      // Remove tier from ID
                    condensedRow[1] = condensedName;    // Remove tier from name
                    condensedRow[5] = '1';              // Set tier to 1

                    condensedRows.push(condensedRow);
                }
            }

            console.log(`\n📊 Condensation Results:`);
            console.log(`Original entries: ${lines.length - 1}`);
            console.log(`Condensed entries: ${condensedRows.length}`);
            console.log(`Reduction: ${((lines.length - 1 - condensedRows.length) / (lines.length - 1) * 100).toFixed(1)}%`);

            // Create backup
            console.log('\n💾 Creating backup...');
            fs.copyFileSync(this.inputFile, this.backupFile);

            // Write condensed file
            console.log('✏️ Writing condensed CSV...');
            const condensedContent = [
                this.arrayToCSVRow(header),
                ...condensedRows.map(row => this.arrayToCSVRow(row))
            ].join('\n');

            fs.writeFileSync(this.outputFile, condensedContent);

            console.log(`\n✅ Condensation complete!`);
            console.log(`📁 Backup saved as: ${this.backupFile}`);
            console.log(`📁 Condensed file saved as: ${this.outputFile}`);

            // Show some sample entries
            console.log('\n📋 Sample condensed entries:');
            for (let i = 0; i < Math.min(5, condensedRows.length); i++) {
                const row = condensedRows[i];
                console.log(`  ${row[0]} - ${row[1]} (${row[4]})`);
            }

            return true;

        } catch (error) {
            console.error('❌ Error during condensation:', error.message);
            return false;
        }
    }

    // Analyze current structure
    analyzeStructure() {
        try {
            console.log('🔍 Analyzing current CSV structure...');
            const content = fs.readFileSync(this.inputFile, 'utf8');
            const lines = content.split('\n').filter(line => line.trim());

            const tierCounts = {};
            const componentTypes = {};
            const sizes = {};

            for (let i = 1; i < lines.length; i++) {
                const row = this.parseCSVRow(lines[i]);

                if (row.length < 6) continue;

                const outputType = row[2];
                const size = row[4];
                const tier = parseInt(row[5]) || 1;

                tierCounts[tier] = (tierCounts[tier] || 0) + 1;
                componentTypes[outputType] = (componentTypes[outputType] || 0) + 1;
                sizes[size] = (sizes[size] || 0) + 1;
            }

            console.log('📊 Current structure analysis:');
            console.log('Tier distribution:', tierCounts);
            console.log('Component types:', Object.keys(componentTypes).length);
            console.log('Ship sizes:', Object.keys(sizes));

            return {
                totalEntries: lines.length - 1,
                tierCounts,
                componentTypes,
                sizes
            };

        } catch (error) {
            console.error('❌ Error analyzing structure:', error.message);
            return null;
        }
    }
}

// Main execution
console.log('🚀 Starting CSV Condensation Process...');
console.log('==========================================');

const condenser = new CSVCondenser();

// Analyze current structure
const analysis = condenser.analyzeStructure();

if (analysis) {
    console.log('\n🔄 Proceeding with condensation...');
    console.log('==========================================');

    const success = condenser.condenseCSV();

    if (success) {
        console.log('\n🎉 CSV condensation completed successfully!');
        console.log('\nThe condensed file contains only the base recipes for each component size,');
        console.log('removing the tiered versions (T1-T5) while maintaining all other data.');
    } else {
        console.log('\n❌ CSV condensation failed. Please check the error messages above.');
    }
} else {
    console.log('\n❌ Could not analyze CSV structure. Please check the file exists and is readable.');
} 

File: public/condense_csv.py
#!/usr/bin/env python3
"""
CSV Condenser Script
Removes redundant T2-T5 components from WorkingCopy.csv, keeping only T1 versions.

The T2-T5 versions are predictable since they only add the previous tier as the first ingredient
while keeping all other ingredients the same.
"""

import csv
import os

def condense_csv(input_file, output_file=None):
    """
    Condense the CSV by keeping only T1 components (OutputTier = 1)
    
    Args:
        input_file (str): Path to the input CSV file
        output_file (str): Path to the output CSV file (if None, overwrites input)
    """
    if output_file is None:
        output_file = input_file
    
    # Read the original CSV
    t1_rows = []
    header = None
    total_rows = 0
    t1_count = 0
    
    print(f"Reading {input_file}...")
    
    with open(input_file, 'r', newline='', encoding='utf-8') as infile:
        reader = csv.reader(infile)
        
        # Read header
        header = next(reader)
        t1_rows.append(header)
        
        # Process each row
        for row in reader:
            total_rows += 1
            
            # Check if this is a T1 component (OutputTier column is index 5)
            if len(row) > 5 and row[5] == '1':
                t1_rows.append(row)
                t1_count += 1
            
            # Print progress every 1000 rows
            if total_rows % 1000 == 0:
                print(f"Processed {total_rows} rows, found {t1_count} T1 components...")
    
    print(f"\nOriginal file: {total_rows} rows")
    print(f"T1 components: {t1_count}")
    print(f"Rows removed: {total_rows - t1_count}")
    print(f"Reduction: {((total_rows - t1_count) / total_rows * 100):.1f}%")
    
    # Write the condensed CSV
    print(f"\nWriting condensed CSV to {output_file}...")
    
    with open(output_file, 'w', newline='', encoding='utf-8') as outfile:
        writer = csv.writer(outfile)
        writer.writerows(t1_rows)
    
    print(f"Condensed CSV saved with {len(t1_rows)} rows (including header)")
    
    return t1_count, total_rows

def main():
    """Main function to run the script"""
    input_file = 'WorkingCopy.csv'
    
    # Check if file exists
    if not os.path.exists(input_file):
        print(f"Error: {input_file} not found!")
        print("Please run this script from the directory containing WorkingCopy.csv")
        return
    
    # Create backup
    backup_file = 'WorkingCopy_backup.csv'
    print(f"Creating backup: {backup_file}")
    
    with open(input_file, 'r', encoding='utf-8') as src:
        with open(backup_file, 'w', encoding='utf-8') as dst:
            dst.write(src.read())
    
    # Condense the CSV
    try:
        t1_count, total_rows = condense_csv(input_file)
        
        print(f"\n✅ Success! CSV condensed from {total_rows} to {t1_count} components")
        print(f"📁 Backup saved as: {backup_file}")
        print(f"📄 Condensed file: {input_file}")
        
        print(f"\n📊 Summary:")
        print(f"   • Original components: {total_rows}")
        print(f"   • T1 components kept: {t1_count}")
        print(f"   • T2-T5 components removed: {total_rows - t1_count}")
        print(f"   • File size reduction: ~{((total_rows - t1_count) / total_rows * 100):.1f}%")
        
    except Exception as e:
        print(f"❌ Error: {e}")
        print("Restoring from backup...")
        
        # Restore backup if something went wrong
        with open(backup_file, 'r', encoding='utf-8') as src:
            with open(input_file, 'w', encoding='utf-8') as dst:
                dst.write(src.read())
        
        print("Original file restored.")

if __name__ == "__main__":
    main() 